### PR TITLE
Add logrotate example to backend reference

### LINKED ISSUE #2381: Add Logrotate content
### Description

Add information on using logrotate to rotate log files when using `event-log-file`

### Category

This is an issue with:

- [ ] errata (error in doc content)
- [ ] new doc (requires a completely new doc)
- [x] update (add missing or refresh existing content)
- [ ] improvement (improve docs content or format)

### Affected pages

https://docs.sensu.io/sensu-go/latest/reference/backend/#event-logging
### Corrections needed

include an example log rotate configuration:

```
/var/log/sensu/events.log {
  rotate 7
  missingok
  notifempty
  compress
  delaycompress
  size 30k
  daily
  create 0644 sensu sensu
  postrotate
  /sbin/service sensu-backend reload > /dev/null 2>/dev/null || true
  endscript
}
```

The important part to point out is the postrotate script which will reload the backend after log rotate is complete. Not doing so will cause sensu-backend (and agent if translated for agent..) to no longer write to the file even if recreated by logrotate. 

### Additional context

* Browser(s) and version(s): 
* OS:

### PR DESCRIPTION
**Pull request by hillaryfraley**
## Description
Adds logrotate example and postrotate script explanation to event logging examples in backend reference.

## Motivation and Context
Closes https://github.com/sensu/sensu-docs/issues/2381

## Review Instructions
I don't think my explanation is sufficient to make it clear how the postrotate example is different from the systemd example that comes right after it. It looks like the postrotate script is slightly different, but I'm not sure how to explain the significance of those differences.

### CODE REVIEW

**Comment by rgeniesse** [comment]
This probably even relates to https://github.com/sensu/sensu-docs/issues/1533. So I can help out here as I've been sitting on that issue for far too long. We may also want to consider if having a guide is required still or having these in the backend/agent reference is more discoverable.

**Comment by hillaryfraley** [comment]
Thanks, @rgeniesse . Can you help me explain how the logrotate and systemd examples are different? To me, it just looks like they save in a different place, but I am not sure if that's correct.

**Comment by rgeniesse** [comment]
@hillaryfraley that appears to be the only difference. The reason the guide is named `systemd-logs` is that systems based off that with Sensu will no log to disk by default, but rather only to `journald`. All that said, I can't say I have tested Sensu packages against non-systemd based systems (those without journald) to know if logging to disk is done.

**Comment by hillaryfraley** [comment]
@rgeniesse Thank you. I added a couple clarifications. I do think it is confusing to have this info in two places, especially since the `postrotate` script is so different in the guide and the reference.

I'm not sure why the script is so different between the guide and the reference...is that something we need to correct?

Maybe the log rotation examples from https://docs.sensu.io/sensu-go/latest/reference/backend/#log-rotation should be moved to the end of the guide (https://docs.sensu.io/sensu-go/latest/guides/systemd-logs/), and a link to the guide added at the top of the Event Logging section (https://docs.sensu.io/sensu-go/latest/reference/backend/#event-logging). I think the event logging flag tables should stay in the backend reference.

@Blue0ctober what do you think?

**Comment by rgeniesse** [comment]
Also, I guess I wasn't aware that this content already existed in the backend reference. Currently it only describes how to get Sensu logging added if you already have a log rotation in place. The guide that the issue I linked to goes through the entire process from the start for systemd. I am not sure how to real in all this content, but at the very lease our rotation configuration should be the same if we keep both places.

**Comment by hillaryfraley** [comment]
Use systemd example with explanation from new example.
Replace sysvinit example script with new example script
Add link to backend reference in log guide

**Comment by hillaryfraley** [comment]
I think this is ready for review now.

@rgeniesse do you think these changes address the issue in #1533 , specifically this comment: https://github.com/sensu/sensu-docs/issues/1533#issuecomment-501581591?

File: content/sensu-go/5.19/reference/backend.md [hunk 1]
```diff
@@ -1243,12 +1244,13 @@ event-log-file: "/var/log/sensu/events.log"{{< /highlight >}}
 ### Log rotation
 
 Event logging supports log rotation via the _SIGHUP_ signal.
-First, rename (move) the current log file.
+To rotate event logs, first rename (move) the current log file.
```
Comment (cwjohnston) [inline content/sensu-go/5.19/reference/backend.md]:
```suggestion
To manually rotate event logs, first rename (move) the current log file.
```

File: content/sensu-go/5.19/reference/backend.md [hunk 2]
```diff
@@ -1243,12 +1244,13 @@ event-log-file: "/var/log/sensu/events.log"{{< /highlight >}}
...
 Then, send the _SIGHUP_ signal to the sensu-backend process so it creates a new log file and starts logging to it.
 
-Here are some log rotate sample configurations:
+#### Log rotation for systemd logs
```
Comment (cwjohnston) [inline content/sensu-go/5.19/reference/backend.md]:
```suggestion
Because event log files can grow very quickly for larger Sensu installations, we recommend using `logrotate` to automatically rotate log files.

#### Log rotation for systemd
```

File: content/sensu-go/5.19/reference/backend.md [hunk 3]
```diff
@@ -1263,7 +1265,13 @@ Here are some log rotate sample configurations:
 }
 {{< /highlight >}}
 
-#### sysvinit
+Without the `postrotate` script, the backend will not reload.
+This will cause sensu-backend (and sensu-agent, if translated for the Sensu agent) to no longer write to the log file, even if logrotate recreates the log file.
+
+Read [Log Sensu services with systemd][28] to learn how to add log forwarding from journald to syslog, have rsyslog write logging data to disk, and set up log rotation of the newly created log files.
+
+#### Log rotation for sysvinit logs
```
Comment (cwjohnston) [inline content/sensu-go/5.19/reference/backend.md]:
```suggestion
#### Log rotation for sysvinit
```

File: content/sensu-go/5.19/guides/systemd-logs.md [hunk 4]
```diff
@@ -84,6 +88,9 @@ The `-d` flag will output details, but it will not take action on the logs or ex
 logrotate -d /etc/logrotate.d/sensu.conf
 {{< /highlight>}}
 
-{{% notice note %}}
-**NOTE**: On Ubuntu systems, be sure to run `chown -R syslog:adm /var/log/sensu` so syslog can write to that directory.
-{{% /notice %}}
+Read the [Sensu backend reference][1] for more information about event logging and log rotation, including [examples for systemd and sysvinit][2].
```
Comment (cwjohnston) [inline content/sensu-go/5.19/guides/systemd-logs.md]:
```suggestion
Sensu also offers logging of event data to a separate log file as a commercial feature. See [Sensu backend reference][1] for more information about event logging.
```

File: content/sensu-go/5.19/guides/systemd-logs.md [hunk 5]
```diff
@@ -84,6 +88,9 @@ The `-d` flag will output details, but it will not take action on the logs or ex
...
+
+
+[1]: ../../reference/backend/#event-logging
+[2]: ../../reference/backend/#log-rotation
```
Comment (cwjohnston) [inline content/sensu-go/5.19/guides/systemd-logs.md]:
```suggestion
```

File: content/sensu-go/5.19/reference/backend.md [hunk 6]
```diff
@@ -1263,7 +1265,13 @@ Here are some log rotate sample configurations:
 }
 {{< /highlight >}}
 
-#### sysvinit
+Without the `postrotate` script, the backend will not reload.
+This will cause sensu-backend (and sensu-agent, if translated for the Sensu agent) to no longer write to the log file, even if logrotate recreates the log file.
+
+Read [Log Sensu services with systemd][28] to learn how to add log forwarding from journald to syslog, have rsyslog write logging data to disk, and set up log rotation of the newly created log files.
```
Comment (cwjohnston) [inline content/sensu-go/5.19/reference/backend.md]:
```suggestion
{{% notice note %}}
**NOTE**: Event logs do not include log messages produced by sensu-backend service. If you wish to write Sensu service logs to flat files on disk, read [Log Sensu services with systemd][28] to learn how.
{{% /notice %}}
```

**Review by cwjohnston — CHANGES_REQUESTED** [review]
A few suggestions.

File: content/sensu-go/5.19/guides/systemd-logs.md [hunk 7]
```diff
@@ -84,6 +88,8 @@ The `-d` flag will output details, but it will not take action on the logs or ex
 logrotate -d /etc/logrotate.d/sensu.conf
 {{< /highlight>}}
 
-{{% notice note %}}
-**NOTE**: On Ubuntu systems, be sure to run `chown -R syslog:adm /var/log/sensu` so syslog can write to that directory.
-{{% /notice %}}
+Sensu also offers logging of event data to a separate log file as a commercial feature. See [Sensu backend reference][1] for more information about event logging.
```
Comment (hillaryfraley) [inline content/sensu-go/5.19/guides/systemd-logs.md]:
```suggestion
Sensu also offers logging of event data to a separate log file as a [commercial feature][2].
See the [Sensu backend reference][1] for more information about event logging.
```

File: content/sensu-go/5.19/guides/systemd-logs.md [hunk 8]
```diff
@@ -84,6 +88,7 @@ The `-d` flag will output details, but it will not take action on the logs or ex
...
+
+
+[1]: ../../reference/backend/#event-logging
```
Comment (hillaryfraley) [inline content/sensu-go/5.19/guides/systemd-logs.md:95]:
```suggestion
[1]: ../../reference/backend/#event-logging
[2]: ../../getting-started/enterprise/
```

File: content/sensu-go/5.19/reference/backend.md [hunk 9]
```diff
@@ -1263,7 +1267,15 @@ Here are some log rotate sample configurations:
 }
 {{< /highlight >}}
 
-#### sysvinit
+Without the `postrotate` script, the backend will not reload.
+This will cause sensu-backend (and sensu-agent, if translated for the Sensu agent) to no longer write to the log file, even if logrotate recreates the log file.
+
+{{% notice note %}}
+**NOTE**: Event logs do not include log messages produced by sensu-backend service. If you wish to write Sensu service logs to flat files on disk, read [Log Sensu services with systemd][28] to learn how.
```
Comment (hillaryfraley) [inline content/sensu-go/5.19/reference/backend.md]:
```suggestion
**NOTE**: Event logs do not include log messages produced by sensu-backend service. To write Sensu service logs to flat files on disk, read [Log Sensu services with systemd][28].
```

File: content/sensu-go/5.19/reference/backend.md [hunk 10]
```diff
@@ -1243,12 +1244,15 @@ event-log-file: "/var/log/sensu/events.log"{{< /highlight >}}
 ### Log rotation
 
 Event logging supports log rotation via the _SIGHUP_ signal.
-First, rename (move) the current log file.
+To manually rotate event logs, first rename (move) the current log file.
 Then, send the _SIGHUP_ signal to the sensu-backend process so it creates a new log file and starts logging to it.
 
-Here are some log rotate sample configurations:
+Because event log files can grow very quickly for larger Sensu installations, we recommend using `logrotate` to automatically rotate log files.
```
Comment (cwjohnston) [inline content/sensu-go/5.19/reference/backend.md]: Most linux distributions include `logrotate` as a standard utility, but by default it is configured to run once a day. As a result, folks who want to use our example logrotate configuration to rotate logs more frequently than once a day may need to take steps to configure logrotate to run once an hour [as described here](https://unix.stackexchange.com/questions/29574/how-can-i-set-up-logrotate-to-rotate-logs-hourly).

File: content/sensu-go/5.19/reference/backend.md [hunk 11]
```diff
@@ -1273,7 +1285,7 @@ Here are some log rotate sample configurations:
   notifempty
   compress
   postrotate
-    kill -HUP `cat /var/run/sensu/sensu-backend.pid 2> /dev/null` 2> /dev/null || true
+    /sbin/service sensu-backend reload > /dev/null 2>/dev/null || true
```
Comment (cwjohnston) [inline content/sensu-go/5.19/reference/backend.md]: I think we need to revert the change to this line -- I've had a look at the sysvinit scripts and they don't implement the logic needed for `service sensu-backend reload` to work. The `kill -HUP` approach is correct for sysvinit.

File: content/sensu-go/5.19/reference/backend.md [hunk 12]
```diff
@@ -1263,7 +1267,15 @@ Here are some log rotate sample configurations:
 }
 {{< /highlight >}}
 
-#### sysvinit
+Without the `postrotate` script, the backend will not reload.
+This will cause sensu-backend (and sensu-agent, if translated for the Sensu agent) to no longer write to the log file, even if logrotate recreates the log file.
```
Comment (cwjohnston) [inline content/sensu-go/5.19/reference/backend.md:1330]: Per [Eric's comment here](https://github.com/sensu/sensu-enterprise-go/issues/1010), I think we need to make some adjustments or add a call out. 

Using `systemctl reload` sends a SIGHUP to the sensu-backend process.

Prior to 5.19.3, a reload will cause sensu-backend to fully restart -- so operators should be mindful of this so that they don't simultaneously restart all the members of a clustered backend. 

For version 5.19.3 and later, SIGHUP will cause only event logging components to reload. For backends running 5.19.3, it should be safe for all members of a backend cluster to receive a SIGHUP simultaneously.

**Review by cwjohnston — CHANGES_REQUESTED** [review]

File: content/sensu-go/5.19/reference/backend.md (hunk 11 above)
Comment (hillaryfraley) [inline content/sensu-go/5.19/reference/backend.md]:
```suggestion
    kill -HUP `cat /var/run/sensu/sensu-backend.pid 2> /dev/null` 2> /dev/null || true
```

**Comment by cwjohnston** [comment]
> @rgeniesse do you think these changes address the issue in #1533 , specifically this comment: [#1533 (comment)](https://github.com/sensu/sensu-docs/issues/1533#issuecomment-501581591)?

I think the rsyslog config examples in guides/systemd-logs.md will need changes in order to address the user comment. As I understand it, they are seeing Sensu log messages in both the intended service log (e.g. `/var/log/sensu/sensu-agent.log` ) *and* their general syslog (e.g. ` /var/log/syslog`). I am seeing the same in my test environment.

File: content/sensu-go/5.19/reference/backend.md (hunk 10 above)
Comment (hillaryfraley) [inline content/sensu-go/5.19/reference/backend.md]:
```suggestion
Most Linux distributions include `logrotate` to automatically rotate log files as a standard utility, configured to run once per day by default.
Because event log files can grow quickly for larger Sensu installations, we recommend using `logrotate` to automatically rotate log files more frequently.
To use the example log rotation configurations listed below, you may need to [configure `logrotate` to run once per hour][29].
```

File: content/sensu-go/5.19/reference/backend.md [hunk 13]
```diff
@@ -1306,3 +1318,4 @@ Here are some log rotate sample configurations:
 [25]: ../../installation/install-sensu#3-initialize
 [26]: ../../sensuctl/reference/#change-admin-user-s-password
 [27]: #configuration-via-environment-variables
+[28]: ../../guides/systemd-logs/
```
Comment (hillaryfraley) [inline content/sensu-go/5.19/reference/backend.md:1385]:
```suggestion
[28]: ../../guides/systemd-logs/
[29]: https://unix.stackexchange.com/questions/29574/how-can-i-set-up-logrotate-to-rotate-logs-hourly
```

File: content/sensu-go/5.19/reference/backend.md [hunk 14]
```diff
@@ -1306,3 +1320,5 @@ Here are some log rotate sample configurations:
...
+[29]: https://unix.stackexchange.com/questions/29574/how-can-i-set-up-logrotate-to-rotate-logs-hourly
```
Comment (hillaryfraley) [inline content/sensu-go/5.19/reference/backend.md:1386]:
```suggestion
[29]: https://unix.stackexchange.com/questions/29574/how-can-i-set-up-logrotate-to-rotate-logs-hourly
[30]: ../../release-notes/#5-19-3-release-notes
```

File: content/sensu-go/5.19/reference/backend.md [hunk 15]
```diff
@@ -1263,7 +1269,15 @@ Here are some log rotate sample configurations:
 }
 {{< /highlight >}}
 
-#### sysvinit
+Without the `postrotate` script, the backend will not reload.
+This will cause sensu-backend (and sensu-agent, if translated for the Sensu agent) to no longer write to the log file, even if logrotate recreates the log file.
```
Comment (hillaryfraley) [inline content/sensu-go/5.19/reference/backend.md:1330]:
```suggestion
This will cause sensu-backend (and sensu-agent, if translated for the Sensu agent) to no longer write to the log file, even if logrotate recreates the log file.

In this script, `systemctl reload` sends a _SIGHUP_ signal to the sensu-backend process.

- In Sensu Go versions 5.19.2 and earlier, this reload causes sensu-backend to fully restart.
If you are running a clustered backend, all cluster members will restart at the same time.
- In [Sensu Go version 5.19.3][30] and later, the _SIGHUP_ signal causes only event logging components to reload, so cluster members will not simultaneously restart.
```

File: content/sensu-go/5.19/reference/backend.md [hunk 16]
```diff
@@ -1243,12 +1244,17 @@ event-log-file: "/var/log/sensu/events.log"{{< /highlight >}}
 ### Log rotation
 
 Event logging supports log rotation via the _SIGHUP_ signal.
-First, rename (move) the current log file.
+To manually rotate event logs, first rename (move) the current log file.
 Then, send the _SIGHUP_ signal to the sensu-backend process so it creates a new log file and starts logging to it.
```
Comment (hillaryfraley) [inline content/sensu-go/5.19/reference/backend.md:1305]: @cwjohnston seems like this might be the right place to mention the reload issue with clustered backends prior to 5.19.3, but I'm not sure if that only applies to `systemctl reload` or if it would happen for the sysvinit example too.

Comment (cwjohnston) [reply content/sensu-go/5.19/reference/backend.md:1305]: Unfortunately I believe it applies equally to sysvinit and systemd. Sending `SIGHUP` manually, `systemctl reload sensu-backend` and /etc/init.d/sensu-backend reload` are all doing the same thing.

File: content/sensu-go/5.19/reference/backend.md (hunk 12 above)
Comment (hillaryfraley) [reply content/sensu-go/5.19/reference/backend.md:1301]: @cwjohnston I think I'm confused because this comment only mentions `systemctl reload`, which is in our systemd example above but is not in our sysvinit example here.

The postrotate script in our sysvinit example does not contain `systemctl reload` or even mention `reload` at all:
`kill -HUP `cat /var/run/sensu/sensu-backend.pid 2> /dev/null` 2> /dev/null || true`

Does this sysvinit script result in the same reload behavior you describe for `systemctl reload`?

File: content/sensu-go/5.19/reference/backend.md [hunk 17]
```diff
@@ -1291,7 +1297,21 @@ Here are some log rotate sample configurations:
...
 
-#### sysvinit
+Without the `postrotate` script, the backend will not reload.
+This will cause sensu-backend (and sensu-agent, if translated for the Sensu agent) to no longer write to the log file, even if logrotate recreates the log file.
+
+In this script, `systemctl reload` sends a _SIGHUP_ signal to the sensu-backend process.
+
+- In Sensu Go versions 5.19.2 and earlier, this reload causes sensu-backend to fully restart.
+If you are running a clustered backend, all cluster members will restart at the same time.
+- In [Sensu Go version 5.19.3][30] and later, the _SIGHUP_ signal causes only event logging components to reload, so cluster members will not simultaneously restart.
```
Comment (cwjohnston) [inline content/sensu-go/5.19/reference/backend.md]: From reading over https://github.com/sensu/sensu-go/pull/3723 I think the entire `backend` component is reloaded, not just event logging.

```suggestion
- In [Sensu Go version 5.19.3][30] and later, the _SIGHUP_ signal causes the `backend` component to reload instead of restarting the process.
```

That said, we added a handler for SIGUSR1 in 5.20.0. I think we'd be well served to add a section in this doc which describes signal handling and then refer to that section from this log rotate content as needed.

File: content/sensu-go/5.19/reference/backend.md [hunk 18]
```diff
@@ -1291,7 +1297,21 @@ Here are some log rotate sample configurations:
...
 {{< /highlight >}}
 
-#### sysvinit
+Without the `postrotate` script, the backend will not reload.
+This will cause sensu-backend (and sensu-agent, if translated for the Sensu agent) to no longer write to the log file, even if logrotate recreates the log file.
+
+In this script, `systemctl reload` sends a _SIGHUP_ signal to the sensu-backend process.
+
+- In Sensu Go versions 5.19.2 and earlier, this reload causes sensu-backend to fully restart.
+If you are running a clustered backend, all cluster members will restart at the same time.
```
Comment (cwjohnston) [inline content/sensu-go/5.19/reference/backend.md]:
```suggestion
If you are running a clustered backend, rotating logs on all cluster members simultaneously could lead to a service interruption.
```

**Review by cwjohnston — CHANGES_REQUESTED** [review]

File: content/sensu-go/5.19/reference/backend.md (hunk 17 above)
Comment (hillaryfraley) [reply content/sensu-go/5.19/reference/backend.md]: @cwjohnston I am thinking that the best way to do this is:
1. Propagate the changes in this PR to all versions (with minor adjustments based on versions) and merge
2. Create a new issue to explain the signal handler (https://github.com/sensu/sensu-go/pull/3723) in the backend reference and add a reference link in the 5.20 docs.

What do you think?

**Comment by hillaryfraley** [comment]
May need a note for older versions about SIGHUP -- check note about 5.19.3 to see if it covers it.

**Review by cwjohnston — APPROVED** [review]
